plugin/evm/vm_test.go

```diff
@@ -3333,6 +3335,134 @@ func TestMessageSignatureRequestsToVM(t *testing.T) {
+		testNetworkID,
+		testCChainID,
+		addressedPayload.Bytes(),
+	)
+	require.NoError(err)
+	msg1 := hexutil.Bytes(unsignedMessage1.Bytes())
+
+	// valid message 2
+	payloadData2 := []byte{4, 5, 6}
+	addressedPayload2, err := payload.NewAddressedCall(
+		sourceAddress.Bytes(),
+		payloadData2,
+	)
+	require.NoError(err)
+	unsignedMessage2, err := avalancheWarp.NewUnsignedMessage(
+		testNetworkID,
+		testCChainID,
+		addressedPayload2.Bytes(),
+	)
+	require.NoError(err)
+	msg2 := hexutil.Bytes(unsignedMessage2.Bytes())
+
+	// invalid payload warp message
+	unsignedMessage3, err := avalancheWarp.NewUnsignedMessage(
+		testNetworkID,
+		testCChainID,
+		[]byte("invalid payload"),
+	)
+	require.NoError(err)
+	invalidMsgPayload := hexutil.Bytes(unsignedMessage3.Bytes())
+
+	type OffChainWarpMessageTest struct {
+		messages    []hexutil.Bytes
+		expectedErr error
+	}
+	tests := map[string]OffChainWarpMessageTest{
+		"invalid message": {
+			messages:    []hexutil.Bytes{invalidMsg},
+			expectedErr: warp.ErrInvalidWarpMsg,
+		},
+		"invalid payload": {
+			messages:    []hexutil.Bytes{invalidMsgPayload},
+			expectedErr: warp.ErrInvalidWarpMsgPayload,
+		},
+		"valid message": {
+			messages:    []hexutil.Bytes{msg1},
+			expectedErr: nil,
+		},
+		"multiple valid message": {
+			messages:    []hexutil.Bytes{msg2, msg1},
+			expectedErr: nil,
+		},
+		"multiple valid/invalid message": {
+			messages:    []hexutil.Bytes{msg1, invalidMsg},
+			expectedErr: warp.ErrInvalidWarpMsg,
+		},
+		"multiple invalid payload/valid message": {
+			messages:    []hexutil.Bytes{invalidMsgPayload, msg1},
+			expectedErr: warp.ErrInvalidWarpMsgPayload,
+		},
+	}
+
+	testOffChainWarpMessages := func(test OffChainWarpMessageTest) {
+		vm := &VM{}
+		ctx, dbManager, genesisBytes, issuer, _ := setupGenesis(t, genesisJSONSubnetEVM)
+		config := map[string][]hexutil.Bytes{
+			"off-chain-warp-messages": test.messages,
+		}
+
+		configJSON, err := json.Marshal(config)
+		require.NoError(err)
+
+		err = vm.Initialize(
+			context.Background(),
+			ctx,
+			dbManager,
+			genesisBytes,
+			[]byte(""),
+			configJSON,
+			issuer,
+			[]*commonEng.Fx{},
+			nil,
+		)
+		require.ErrorIs(err, test.expectedErr)
+		if test.expectedErr != nil {
+			return
+		}
+
+		defer func() {
+			err := vm.Shutdown(context.Background())
+			require.NoError(err)
+		}()
+
+		// Check that the off-chain warp messages are retrievable
+		for _, message := range test.messages {
+			unsignedWarpMessage, err := avalancheWarp.ParseUnsignedMessage(message)
+			require.NoError(err)
+			expectedSignature, err := vm.ctx.WarpSigner.Sign(unsignedWarpMessage)
+			require.NoError(err)
+
+			// On warp backend initialization off-chain warp messages were put into the off-chain message map
+			signature, err := vm.warpBackend.GetMessageSignature(unsignedWarpMessage.ID())
+			require.NoError(err)
+			require.Equal(([bls.SignatureLen]byte)(expectedSignature), signature)
+		}
+	}
```

Comment:
Since block payloads are served from a different handler, we need to parse the payload and add a type switch to check which of the two functions it should call. Additionally, calling `GetMessageSignature` on a `blockID` of a block hash payload should not return a signature.